miditok/midi_tokenizer.py

```diff
@@ -1136,54 +1137,26 @@ def __create_time_signatures(self) -> List[Tuple]:
-        ), "The beat resolution in time signature must be a power of 2"
+        time_signature_range = self.config.time_signature_range
 
         time_signatures = []
-        for i in range(0, int(math.log2(max_beat_res)) + 1):  # 1 ~ max_beat_res
-            for j in range(1, ((2**i) * nb_notes) + 1):
-                time_signatures.append((j, 2**i))
-        return time_signatures
-
-    def _reduce_time_signature(
-        self, numerator: int, denominator: int
-    ) -> Tuple[int, int]:
-        r"""Reduces and decomposes a time signature into one of the valid vocabulary time signatures.
-        If time signature's denominator (beat resolution) is larger than max_beat_res,
-        the denominator and numerator are reduced to max_beat_res if possible.
-        If time signature's numerator (bar length in beats) is larger than nb_notes * denominator,
-        the numerator is replaced with its GCD not larger than nb_notes * denominator.
-
-        Example: (10, 4), max_beat_res of 8, and nb_notes of 2 will convert the signature into (5, 4).
-
-        :param numerator: time signature's numerator (bar length in beats).
-        :param denominator: time signature's denominator (beat resolution).
-        :return: the numerator and denominator of a reduced and decomposed time signature.
-        """
-        max_beat_res, nb_notes = self.config.time_signature_range
+        for beat_res, beats in time_signature_range.items():
+            assert beat_res > 0 and math.log2(beat_res).is_integer(), \
+                f"The beat resolution ({beat_res}) in time signature must be a power of 2"
 
-        # reduction (when denominator exceed max_beat_res)
-        while (
-            denominator > max_beat_res and denominator % 2 == 0 and numerator % 2 == 0
-        ):
-            denominator //= 2
-            numerator //= 2
+            time_signatures.extend([(nb_beats, beat_res) for nb_beats in beats])
 
-        assert denominator <= max_beat_res, (
-            f"Unsupported time signature ({numerator}/{denominator}), "
-            f"beat resolution is irreducible to maximum beat resolution {max_beat_res}"
-        )
+        return time_signatures
 
-        # decomposition (when length of a bar exceed max_nb_beats_per_bar)
-        while numerator > nb_notes * denominator:
-            for i in range(2, numerator + 1):
-                if numerator % i == 0:
-                    numerator //= i
-                    break
+    @staticmethod
+    def _compute_ticks_per_bar(time_sig: TimeSignature, time_division: int):
+        r"""Computes time resolution of one bar in ticks.
 
-        return numerator, denominator
+        :param time_sig: time signature object
+        :param time_division: MIDI time division / resolution, in ticks/beat (of the MIDI being parsed)
+        :return: MIDI bar resolution, in ticks/bar
+        """
+        return int(time_division * 4 * time_sig.numerator / time_sig.denominator)
```

Comment:
I came to this empirically when I was working with MIDI files with multiple time signatures, computing bar boundaries, and looking at note start tick times. Also MusicBERT (Octuple) uses the same [formula](https://github.com/microsoft/muzic/blob/main/musicbert/preprocess.py#L202).

From what I could find now, it has nothing to do with MIDI playback, but can be used to compute exact bar numbers and boundaries, which we need in encodings that use Bar tokens.

Some links:
https://music.stackexchange.com/questions/36323/whats-the-purpose-of-time-signature-information-in-a-midi-file
https://stackoverflow.com/questions/61298392/time-signature-meta-message-in-midi
https://www.kvraudio.com/forum/viewtopic.php?t=524002

Reply:
Awesome! I just got it now, I don't now why but my mind was stuck on the time_division expressed in ticks per beat whereas it is actually in ticks per quarter note!

Now that I think of it, all the durations / time shift values are expressed in beats and based in pure ticks / beat, even when encoding / decoding time signatures. If we wanted to be more accurate and represent tokens with more musical sense, we could base their values on ticks / quarter note instead. This would however mean that for two different TSs -with different denum), a specific token values (expressed in beats) would represent different durations in absolute tick.
That's just an idea, what do you think about this ? (I'll resolve this conversation, you can answer anywhere else)
The downsides might be that if there is no recurrent TS tokens within a sequence, so that the model can relate to in order to determine the beat values, the model could struggle to capture the real durations / time shift values in a coherent way. 🤷‍♂️

Reply:
Do you mean encoding durations based on beat value inside each "time signature region"? Durations in ticks scaled by time division are exactly the ratios of the quarter note.

Reply:
Yes that's what I meant, sorry for the confusion.
ticks per beat value is more accurate